plugin/fallback/fallback_test.go

```diff
@@ -0,0 +1,226 @@
+package fallback
+
+import (
+	"context"
+	"testing"
+
+	"github.com/miekg/dns"
+
+	"fmt"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/pkg/healthcheck"
+	"github.com/coredns/coredns/plugin/proxy"
+	"github.com/coredns/coredns/plugin/test"
+	"github.com/coredns/coredns/request"
+)
+
+// stubNextHandler is used to simulate a proxy plugin.
+// It returns a stub Handler that returns the rcode and err specified when invoked.
+func stubNextHandler(rcode int, err error) test.Handler {
+	return test.HandlerFunc(func(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+		w.WriteMsg(&dns.Msg{})
+		return rcode, err
+	})
+}
+
+// testProxyCreator implements the proxyCreator interface.
+// Used by unit test to verify that the proxyCreator.Create() method is called as expected
+type testProxyCreator struct {
+	expectedUpstream proxy.Upstream
+	called           int
+	t                *testing.T
+}
+
+func (c *testProxyCreator) Create(trace plugin.Handler, upstream proxy.Upstream) plugin.Handler {
+	c.called++
+
+	// Ensure that it is called with the expected Upstream
+	if c.expectedUpstream != nil && c.expectedUpstream != upstream {
+		c.t.Fatalf("Expected upstream passed to proxyCreator is '%v', but got '%v'",
```

Comment:
calling t.Fatalf in a non test function? Should return an err

Reply:
As per comment above, the check here is indeed part of the test.

Having said that, I do think it is more appropriate to use t.Errorf (instead of t.Fatalf). Will fix.